### DISCUSSION TITLE
Revert to Zookeeper AFTER Kraft migration

### DISCUSSION THREAD

**Discussion by jonsnowseven** [strimzi/strimzi-kafka-operator#10107] [Q&A] [answered]
Hello!

Is there a guide to rollback to Zookeeper **after** Kraft migration is completed?

I checked that is [possible to rollback while in migration](https://strimzi.io/blog/2024/03/22/strimzi-kraft-migration/#rolling-back-the-cluster-to-zookeeper) but didn't find a walkthrough if the migration is complete.

**Comment by scholzj** [comment]
I don't think you can migrate from KRaft to ZooKeeper. I do not think Apache Kafka supports that.

**Reply by ppatierno** [reply, marked as answer]
No there is no way to do so. When you complete the migration to KRaft, you cannot revert back to ZooKeeper. You can rollback until the migration state is still in `KRaftPostMigration` where ZooKeeper is still up and running and KRaft controllers are in dual-write mode so they are still copying metadata to ZooKeeper. If you end the migration to the `KRaft` state, you are done. No revert back is possible. As Jakub said, it's not about Strimzi. It's about Kafka.

**Reply by jonsnowseven** [reply]
So what happens if I revert back the changes I did in the migration to Kraft (literally reverting the commits)?

**Reply by ppatierno** [reply]
Which commits? 
When you end the migration you have `strimzi.io/kraft: enabled` annotation.
If you try to apply any other value from here (i.e. `disabled`) because you would like to go back to ZooKeeper, it will raise a warning that the annotation value is not allowed in the `KRaft` state and nothing will happen. As I said. There is no way to come back. Apache Kafka doesn't support it at all, leaving Strimzi out of the picture.

**Reply by jonsnowseven** [reply]
I did these [steps](https://strimzi.io/blog/2024/03/22/strimzi-kraft-migration/#rolling-back-the-cluster-to-zookeeper) by committing the changes to the repository where I have the strimzi template.

If I revert these changes (remove the node pools, the annotations), what happens?

**Reply by ppatierno** [reply]
Let's ask a simple question, because you still point me to the steps talking about rollback. Did you end or not the migration? Did you apply the `strimzi.io/kraft: enabled` annotation? Is your `kafkaMetadataState` field in the `Kafka` CR status at `KRaft`? If the answer is yes for my questions, you ended the migration, ZooKeeper is not even running anymore and you are in KRaft, no way to go back, whatever you do. Removing node pools will just break the cluster. Changing the annotation will just raise a warning as I said.

Instead .. if you are still in the `KRaftPostMigration` state as described in the blog post, you can still rollback by applying the `strimzi.io/kraft: rollback` annotation (and then the `disabled` value by following the procedure). It's not about undoing some commits (if you are using GitOps), it's about doing different commits.

**Reply by jonsnowseven** [reply]
> Did you end or not the migration? Did you apply the strimzi.io/kraft: enabled annotation? Is your kafkaMetadataState field in the Kafka CR status at KRaft?

Yes to both

> It's not about undoing some commits (if you are using GitOps), it's about doing different commits.

Yes, I know. I meant that I wanted to revert all the changes (delete the node pools, delete the annotations and bring Zookeeper nodes back).

**Reply by ppatierno** [reply]
> Yes, I know. I meant that I wanted to revert all the changes (delete the node pools, delete the annotations and bring Zookeeper nodes back)

So again ... if you remove the controllers node pools you are going to break the cluster. You cannot spin up ZooKeeper back again by just adding the related section back. The annotation value won't be allowed, brokers won't be restarted to reuse ZooKeeper etc etc.
I am sorry, but your cluster will stay KRaft forever.

**Reply by ppatierno** [reply]
We also have a clear warning in the doc here https://strimzi.io/docs/operators/latest/deploying#proc-deploy-migrate-kraft-str

> WARNING
You can [roll back to using ZooKeeper from this point](https://strimzi.io/docs/operators/latest/deploying#proc-deploy-migrate-kraft-rollback-str). The next step is to enable KRaft. Rollback cannot be performed after enabling KRaft.

**Reply by jonsnowseven** [reply]
There is no problem in keeping KRaft forever. I just wanted to do some testing with Zookeeper again.
Thank you @ppatierno and @scholzj!

**Reply by lenin-joseph** [reply]
Hi @jonsnowseven , I am just asking out of curiosity, were you able to find any workaround for rollback from KRaft to ZK?

**Reply by jonsnowseven** [reply]
Hello @lenin-joseph!
Nope, we kept KRaft.

**Reply by lenin-joseph** [reply]
Thanks @jonsnowseven for the response.

I tried by deleting /controller and /migration on ZK node and recreating the cluster in ZK mode by retaining the PVC, it worked a couple of times(2 min downtime during cluster-recreation), but most of the times I ended up with unrecoverable URP.
